File: reth/tx_processor/src/tx_processor/address_balance_change_calculator.rs
/// Address Balance Change Calculator
/// 
/// This module calculates net balance changes for addresses involved in transactions by:
/// 1. Tracking unknown token movements with contract addresses as keys
/// 2. Tracking known currency movements (ETH, USDC, USDT, DAI, MKR, etc.) with symbols as keys
/// 3. Handling special cases like WETH conversions and bribes
/// 4. Filtering out insignificant state changes based on thresholds
/// 
/// Key Features:
/// - Clean separation: token_net uses addresses, currency_net uses symbols
/// - Automatically recognizes 100+ currencies from DENOM_ADDRESSES
/// - Only applies decimal conversion for tokens with known decimals (no RPC calls)
/// - Tracks both incoming and outgoing movements for tokens and currencies
/// - Handles special addresses (WETH, null, dead addresses)
/// - Identifies significant state changes based on configurable thresholds
/// - Excludes WETH conversions from ETH movements
/// - Tracks bribe payments to known fee recipients
/// - Returns token_net with contract addresses only (unknown tokens)
/// - Returns currency_net with symbols only (tokens in DENOM_ADDRESSES)
/// - No overlap between token_net and currency_net

use super::data_models::events::InternalTransaction;
use super::data_models::balance_changes::{AddressBalanceChange, TokenMovements, TokenMovement};
use crate::utils::to_checksum_address;
use alloy_primitives::{Address, U256};
use eyre::Result;
use lazy_static::lazy_static;
use std::collections::HashMap;
use std::str::FromStr;
use reth_chain_query::FEE_RECIPIENTS;

/// Known token addresses mapped to their symbols
lazy_static! {
    static ref DENOM_ADDRESSES: HashMap<Address, &'static str> = {
        let mut m = HashMap::new();
        // WETH
        m.insert(Address::from([0xC0, 0x2a, 0xaA, 0x39, 0xb2, 0x23, 0xFE, 0x8D, 0x0A, 0x0e, 0x5C, 0x4F, 0x27, 0xeA, 0xD9, 0x08, 0x3C, 0x75, 0x6C, 0xc2]), "WETH");
        // USDC
        m.insert(Address::from([0xA0, 0xb8, 0x69, 0x91, 0xc6, 0x21, 0x8b, 0x36, 0xc1, 0xd1, 0x9D, 0x4a, 0x2e, 0x9E, 0xb0, 0xce, 0x36, 0x06, 0xeB, 0x48]), "USDC");
        // USDT
        m.insert(Address::from([0xdA, 0xC1, 0x7F, 0x95, 0x8D, 0x2e, 0xe5, 0x23, 0xa2, 0x20, 0x62, 0x06, 0x99, 0x45, 0x97, 0xC1, 0x3D, 0x83, 0x1e, 0xc7]), "USDT");
        // DAI
        m.insert(Address::from([0x6B, 0x17, 0x54, 0x74, 0xE8, 0x90, 0x94, 0xC4, 0x4D, 0xa9, 0x8b, 0x95, 0x4E, 0xeD, 0xeA, 0xC4, 0x95, 0x27, 0x1d, 0x0F]), "DAI");
        // Add more known tokens here
        m
    };
}

/// Get the token symbol for a known token address
pub fn get_token_symbol(token_address: &Address) -> Option<&'static str> {
    DENOM_ADDRESSES.get(token_address).copied()
}

/// Token decimals for known currencies
lazy_static! {
    static ref ERC20_TOKEN_DECIMALS: HashMap<&'static str, u8> = {
        let mut m = HashMap::new();
        m.insert("ETH", 18);
        m.insert("WETH", 18);
        m.insert("USDC", 6);
        m.insert("USDT", 6);
        m.insert("DAI", 18);
        // Add more token decimals here
        m
    };
}

/// Type of movement (currency or token)
#[derive(Debug, Clone, Copy)]
enum MovementType {
    Currency,
    Token,
}

/// Unique identifier for a transfer
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
struct TransferId {
    block_number: u64,
    tx_index: u64,
    log_index: String,
}

impl TransferId {
    fn new(block_number: u64, tx_index: u64, log_index: String) -> Self {
        Self {
            block_number,
            tx_index,
            log_index,
        }
    }
}

/// Movement entry for tracking transfers
#[derive(Debug, Clone)]
struct MovementEntry {
    amount: U256,
}

/// Movements for an address (incoming and outgoing)
#[derive(Debug, Clone, Default)]
struct AddressMovements {
    incoming: HashMap<TransferId, MovementEntry>,
    outgoing: HashMap<TransferId, MovementEntry>,
}

/// Balance change calculator
pub struct AddressBalanceChangeCalculator {
    weth_address: Address,
    eth_state_change_threshold: f64,
    token_state_change_threshold: f64,
    // Track movements per address per currency/token
    // currencies[address][currency_symbol] = AddressMovements
    // tokens[address][token_address] = AddressMovements
    currency_movements: HashMap<Address, HashMap<String, AddressMovements>>,
    token_movements: HashMap<Address, HashMap<Address, AddressMovements>>,
}

impl AddressBalanceChangeCalculator {
    /// Create a new balance change calculator
    pub fn new() -> Self {
        Self {
            weth_address: Address::from([0xC0, 0x2a, 0xaA, 0x39, 0xb2, 0x23, 0xFE, 0x8D, 0x0A, 0x0e, 0x5C, 0x4F, 0x27, 0xeA, 0xD9, 0x08, 0x3C, 0x75, 0x6C, 0xc2]),
            eth_state_change_threshold: 0.0005,
            token_state_change_threshold: 0.1,
            currency_movements: HashMap::new(),
            token_movements: HashMap::new(),
        }
    }
    
    /// Calculate balance changes from DECODED transfers and internal transactions
    /// This is the main entry point for tx_processor to calculate balance changes
    pub fn calculate_balance_changes_from_processed_data(
        &mut self,
        erc20_transfers: &[super::data_models::events::ERC20Transfer],
        internal_transactions: &[InternalTransaction],
        block_number: u64,
        tx_index: u64,
    ) -> Result<HashMap<Address, AddressBalanceChange>> {
        // Clear previous state
        self.currency_movements.clear();
        self.token_movements.clear();
        
        // Process internal transactions for ETH transfers
        for (i, internal_tx) in internal_transactions.iter().enumerate() {
            if internal_tx.value > U256::ZERO {
                let transfer_id = TransferId::new(
                    block_number,
                    tx_index,
                    format!("internal_{}", i),
                );
                
                // Keep internal ETH transfers in wei
                let amount_wei = internal_tx.value;
                self.track_movement(
                    MovementType::Currency,
                    internal_tx.from_address,
                    internal_tx.to_address,
                    amount_wei,  // Keep in wei like Python
                    transfer_id,
                    None,
                    Some("ETH".to_string()),
                );
            }
        }
        
        // Process ERC20 transfers (already decoded!)
        for (i, transfer) in erc20_transfers.iter().enumerate() {
            let transfer_id = TransferId::new(
                block_number,
                tx_index,
                transfer.log_index.to_string(),
            );
            
            // Check if this is a known currency (USDC, USDT, DAI, etc.)
            let token_addr = transfer.token_address;
            
            if let Some(mut currency_symbol) = DENOM_ADDRESSES.get(&token_addr).cloned() {
                // EXACT Python logic: Convert WETH to ETH for ERC20
                if currency_symbol == "WETH" {
                    currency_symbol = "ETH";
                }
                
                // Keep amounts as U256 - no decimal conversion needed here
                // The amounts are already in the token's smallest unit
                let amount = transfer.amount;
                
                self.track_movement(
                    MovementType::Currency,
                    transfer.from_address,
                    transfer.to_address,
                    amount,
                    transfer_id,
                    None,
                    Some(currency_symbol.to_string()),
                );
            } else {
                // Unknown token - track with raw amount
                let amount = transfer.amount;
                
                self.track_movement(
                    MovementType::Token,
                    transfer.from_address,
                    transfer.to_address,
                    amount,
                    transfer_id,
                    Some(token_addr),
                    None,
                );
            }
        }
        
        // Calculate final balances - use the from_address if we have transfers
        let from_addr = if !erc20_transfers.is_empty() {
            erc20_transfers[0].from_address
        } else if !internal_transactions.is_empty() {
            internal_transactions[0].from_address
        } else {
            Address::ZERO
        };
        
        self.get_net_balance_changes(from_addr)
    }

    /// Track a movement between addresses and handle special cases
    fn track_movement(
        &mut self,
        movement_type: MovementType,
        from_addr: Address,
        to_addr: Address,
        amount: U256,
        transfer_id: TransferId,
        token_address: Option<Address>,
        currency: Option<String>,
    ) {
        // CORRECTED: Python implementation actually DOES track WETH movements!
        // The filtering should only apply to pure WETH wrap/unwrap operations, 
        // NOT to legitimate DEX trades involving WETH conversions.
        // For now, remove the WETH filtering to match Python behavior.
        // TODO: Implement more sophisticated filtering if needed
        
        // Original filtering (causing bug):
        // if to_addr == self.weth_address || from_addr == self.weth_address {
        //     return;
        // }
        
        match movement_type {
            MovementType::Currency => {
                if let Some(currency_symbol) = currency {
                    // Track per currency (ETH, USDC, USDT, DAI, etc.)
                    // Outgoing from from_addr
                    self.currency_movements
                        .entry(from_addr)
                        .or_default()
                        .entry(currency_symbol.clone())
                        .or_default()
                        .outgoing
                        .insert(transfer_id.clone(), MovementEntry { amount });
                    
                    // Incoming to to_addr (unless it's a bribe to fee recipient for ETH)
                    let is_bribe = currency_symbol == "ETH" && FEE_RECIPIENTS.contains(&to_addr);
                    
                    if !is_bribe {
                        self.currency_movements
                            .entry(to_addr)
                            .or_default()
                            .entry(currency_symbol)
                            .or_default()
                            .incoming
                            .insert(transfer_id, MovementEntry { amount });
                    }
                }
            }
            MovementType::Token => {
                if let Some(token_addr) = token_address {
                    // Track per token address (for unknown tokens not in DENOM_ADDRESSES)
                    
                    // Outgoing from from_addr
                    self.token_movements
                        .entry(from_addr)
                        .or_default()
                        .entry(token_addr)
                        .or_default()
                        .outgoing
                        .insert(transfer_id.clone(), MovementEntry { amount });
                    
                    // Incoming to to_addr
                    self.token_movements
                        .entry(to_addr)
                        .or_default()
                        .entry(token_addr)
                        .or_default()
                        .incoming
                        .insert(transfer_id, MovementEntry { amount });
                }
            }
        }
    }

    /// Aggregate net changes for every address touched in this tx
    fn get_net_balance_changes(&self, from_address: Address) -> Result<HashMap<Address, AddressBalanceChange>> {
        let mut net = HashMap::new();
        
        // Get all addresses that had any movements
        let mut all_addrs: std::collections::HashSet<&Address> = std::collections::HashSet::new();
        all_addrs.extend(self.token_movements.keys());
        all_addrs.extend(self.currency_movements.keys());
        
        for addr in all_addrs {
            // Calculate currency net changes per currency
            let mut currency_net = HashMap::new();
            let mut total_currency_movement = U256::ZERO;
            
            if let Some(currency_movs) = self.currency_movements.get(addr) {
                for (currency_symbol, movements) in currency_movs {
                    let currency_in: U256 = movements.incoming.values().map(|e| e.amount).fold(U256::ZERO, |acc, x| acc + x);
                    let currency_out: U256 = movements.outgoing.values().map(|e| e.amount).fold(U256::ZERO, |acc, x| acc + x);
                    // Calculate net change with proper sign (positive for net gain, negative for net loss)
                    let net_change = if currency_in >= currency_out {
                        currency_in - currency_out  // Positive: net gain
                    } else {
                        // Negative: net loss - use two's complement representation
                        U256::MAX - (currency_out - currency_in) + U256::from(1)
                    };
                    
                    // Apply threshold based on currency type (check absolute value)
                    let threshold = if currency_symbol == "ETH" {
                        // 0.0005 ETH = 500000000000000 wei
                        U256::from(500_000_000_000_000u64)
                    } else {
                        U256::from(self.token_state_change_threshold as u64)
                    };
                    
                    // Check if absolute change exceeds threshold
                    let abs_change = if net_change > U256::MAX / U256::from(2) {
                        // Negative value - get absolute value
                        U256::MAX - net_change + U256::from(1)
                    } else {
                        net_change
                    };
                    
                    if abs_change > threshold {
                        // Store the signed net change
                        currency_net.insert(currency_symbol.clone(), net_change);
                        total_currency_movement = total_currency_movement + abs_change;
                    }
                }
            }
            
            // Calculate token net changes per token (only for tokens NOT in DENOM_ADDRESSES)
            let mut token_net = HashMap::new();
            let mut total_token_movement = U256::ZERO;
            
            if let Some(token_movs) = self.token_movements.get(addr) {
                for (token_addr, movements) in token_movs {
                    // Skip tokens that are in DENOM_ADDRESSES - they go to currency_net
                    if !DENOM_ADDRESSES.contains_key(token_addr) {
                        let token_in: U256 = movements.incoming.values().map(|e| e.amount).fold(U256::ZERO, |acc, x| acc + x);
                        let token_out: U256 = movements.outgoing.values().map(|e| e.amount).fold(U256::ZERO, |acc, x| acc + x);
                        // Calculate net change with proper sign (positive for net gain, negative for net loss)
                        let net_change = if token_in >= token_out {
                            token_in - token_out  // Positive: net gain
                        } else {
                            // Negative: net loss - use two's complement representation
                            U256::MAX - (token_out - token_in) + U256::from(1)
                        };
                        
                        let threshold = U256::from(self.token_state_change_threshold as u64);
                        
                        // Check if absolute change exceeds threshold
                        let abs_change = if net_change > U256::MAX / U256::from(2) {
                            // Negative value - get absolute value
                            U256::MAX - net_change + U256::from(1)
                        } else {
                            net_change
                        };
                        
                        if abs_change > threshold {
                            // Always use checksum address as key, signed amount as value
                            let token_key = to_checksum_address(token_addr);
                            token_net.insert(token_key, net_change);
                            total_token_movement = total_token_movement + abs_change;
                        }
                    }
                }
            }
            
            // Include address if it meets thresholds or is the sender
            if total_token_movement > U256::ZERO
                || total_currency_movement > U256::ZERO
                || *addr == from_address
            {
                // Create proper struct instead of JSON
                let movements = TokenMovements {
                    tokens: self.format_token_movements_struct(addr),
                    currencies: self.format_currency_movements_struct(addr),
                };
                
                let balance_change = AddressBalanceChange {
                    token_net,
                    currency_net,
                    movements,
                };
                
                net.insert(*addr, balance_change);
            }
        }
        
        Ok(net)
    }
    
    /// Format token movements for struct output
    fn format_token_movements_struct(&self, addr: &Address) -> HashMap<String, TokenMovement> {
        let mut result = HashMap::new();
        
        if let Some(token_movs) = self.token_movements.get(addr) {
            for (token_addr, movements) in token_movs {
                let token_key = to_checksum_address(token_addr);
                
                let mut incoming = HashMap::new();
                for (tid, entry) in &movements.incoming {
                    let key = format!("{}_{}_{}",tid.block_number, tid.tx_index, tid.log_index);
                    incoming.insert(key, entry.amount);
                }
                
                let mut outgoing = HashMap::new();
                for (tid, entry) in &movements.outgoing {
                    let key = format!("{}_{}_{}", tid.block_number, tid.tx_index, tid.log_index);
                    outgoing.insert(key, entry.amount);
                }
                
                result.insert(token_key, TokenMovement {
                    incoming,
                    outgoing,
                });
            }
        }
        
        result
    }
    
    /// Format currency movements for struct output
    fn format_currency_movements_struct(&self, addr: &Address) -> HashMap<String, TokenMovement> {
        let mut result = HashMap::new();
        
        if let Some(currency_movs) = self.currency_movements.get(addr) {
            for (currency_symbol, movements) in currency_movs {
                let mut incoming = HashMap::new();
                for (tid, entry) in &movements.incoming {
                    let key = format!("{}_{}_{}",tid.block_number, tid.tx_index, tid.log_index);
                    incoming.insert(key, entry.amount);
                }
                
                let mut outgoing = HashMap::new();
                for (tid, entry) in &movements.outgoing {
                    let key = format!("{}_{}_{}", tid.block_number, tid.tx_index, tid.log_index);
                    outgoing.insert(key, entry.amount);
                }
                
                result.insert(currency_symbol.clone(), TokenMovement {
                    incoming,
                    outgoing,
                });
            }
        }
        
        result
    }
}

impl Default for AddressBalanceChangeCalculator {
    fn default() -> Self {
        Self::new()
    }
}

File: pyeth/eth_data/eth_data/tx_processor/address_balance_change_calculator.py
from collections import defaultdict, OrderedDict
from eth_data.chain_utils.common_addresses import fee_recipients_set, DENOM_ADDRESSES, ERC20_TOKEN_DECIMALS


class AddressBalanceChangeCalculator:
    """
    AddressBalanceChangeCalculator: Tracks and analyzes address balance changes in Ethereum transactions.

    This class calculates net balance changes for addresses involved in transactions by:
    1. Tracking unknown token movements with contract addresses as keys
    2. Tracking known currency movements (ETH, USDC, USDT, DAI, MKR, etc.) with symbols as keys
    3. Handling special cases like WETH conversions and bribes
    4. Filtering out insignificant state changes based on thresholds

    Key Features:
    - Maintains chronological order of transfers using OrderedDict
    - Clean separation: token_net uses addresses, currency_net uses symbols
    - Automatically recognizes 100+ currencies from DENOM_ADDRESSES
    - Only applies decimal conversion for tokens with known decimals (no RPC calls)
    - Tracks both incoming and outgoing movements for tokens and currencies
    - Handles special addresses (WETH, null, dead addresses)
    - Identifies significant state changes based on configurable thresholds
    - Excludes WETH conversions from ETH movements
    - Tracks bribe payments to known fee recipients
    - Returns token_net with contract addresses only (unknown tokens)
    - Returns currency_net with symbols only (tokens in DENOM_ADDRESSES)
    - No overlap between token_net and currency_net
    """

    def __init__(
        self,
        eth_state_change_threshold: float = 0.0005,
        token_state_change_threshold: float = 0.1,
        logger=None,
    ):
        self.logger = logger
        self.WETH_ADDRESS = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
        self.eth_state_change_threshold = eth_state_change_threshold
        self.token_state_change_threshold = token_state_change_threshold
        # Track movements per address per currency/token
        # Structure: 
        #   currencies[address][currency_symbol] = {"in": OrderedDict(), "out": OrderedDict()}
        #   tokens[address][token_address] = {"in": OrderedDict(), "out": OrderedDict()}
        self.movements = {
            "currencies": defaultdict(lambda: defaultdict(lambda: {"in": OrderedDict(), "out": OrderedDict()})),
            "tokens": defaultdict(lambda: defaultdict(lambda: {"in": OrderedDict(), "out": OrderedDict()})),
        }

    def _track_movement(
        self,
        movement_type: str,
        from_addr: str,
        to_addr: str,
        amount: float,
        transfer_id: tuple,
        token_address: str = None,
        currency: str = None,
    ):
        """Track a movement between addresses and handle special cases.
        
        Args:
            movement_type: "currency" or "token"
            from_addr: Source address
            to_addr: Target address
            amount: Amount transferred
            transfer_id: Unique transfer identifier (block, txn_index, log_index)
            token_address: Token contract address (for unknown token movements)
            currency: Currency symbol (ETH, USDC, USDT, etc.) for currency movements
        """
        # WETH conversions are denomination-neutral
        if to_addr == self.WETH_ADDRESS or from_addr == self.WETH_ADDRESS:
            return
        
        if movement_type == "currency" and currency:
            # Track per currency (ETH, USDC, USDT, DAI, etc.)
            # Outgoing from `from_addr`
            self.movements["currencies"][from_addr][currency]["out"][transfer_id] = amount
            # Incoming to `to_addr` (unless it's a bribe to fee recipient for ETH)
            if not (currency == "ETH" and to_addr in fee_recipients_set):
                self.movements["currencies"][to_addr][currency]["in"][transfer_id] = amount
                    
        elif movement_type == "token" and token_address:
            # Track per token address (for unknown tokens not in DENOM_ADDRESSES)
            # Outgoing from `from_addr`
            self.movements["tokens"][from_addr][token_address]["out"][transfer_id] = amount
            # Incoming to `to_addr`
            self.movements["tokens"][to_addr][token_address]["in"][transfer_id] = amount

    def get_net_balance_changes(self, from_address: str) -> dict:
        """
        Aggregate net changes for every address touched in this tx.

        Returns
        -------
        dict[address] -> {
            token_net : dict[token_address] -> float,  # Unknown tokens only, raw amounts
            currency_net : dict[currency_symbol] -> float,  # Known currencies only, decimal-adjusted
            movements : {
                tokens: {token_addr: {in/out}},  # Raw amounts
                currencies: {currency_symbol: {in/out}}  # Decimal-adjusted amounts
            }
        }
        """
        net = {}
        
        # Get all addresses that had any movements
        token_addrs = set(self.movements["tokens"].keys())
        currency_addrs = set(self.movements["currencies"].keys())
        all_addrs = token_addrs | currency_addrs
        
        for addr in all_addrs:
            # Calculate currency net changes per currency
            currency_net = {}
            total_currency_movement = 0.0
            
            for currency_symbol, movements in self.movements["currencies"][addr].items():
                currency_in = sum(movements["in"].values())
                currency_out = sum(movements["out"].values())
                net_change = currency_in - currency_out
                
                # Apply threshold based on currency type
                if currency_symbol == "ETH":
                    # ETH is now in wei, convert threshold to wei for comparison
                    threshold = self.eth_state_change_threshold * 1e18
                else:
                    threshold = self.token_state_change_threshold
                
                if abs(net_change) > threshold:
                    # For display: convert ETH from wei to ether, keep others as is
                    if currency_symbol == "ETH":
                        # Convert wei to ETH for display (but keep original for movements)
                        currency_net[currency_symbol] = net_change / 1e18  # Display in ETH
                    else:
                        currency_net[currency_symbol] = net_change
                    total_currency_movement += abs(net_change)
            
            # Calculate token net changes per token (only for tokens NOT in DENOM_ADDRESSES)
            token_net = {}
            total_token_movement = 0.0
            
            for token_addr, movements in self.movements["tokens"][addr].items():
                # Skip tokens that are in DENOM_ADDRESSES - they go to currency_net
                if token_addr not in DENOM_ADDRESSES:
                    token_in = sum(movements["in"].values())
                    token_out = sum(movements["out"].values())
                    net_change = token_in - token_out
                    
                    if abs(net_change) > self.token_state_change_threshold:
                        # Always use contract address as key, raw amount as value
                        token_net[token_addr] = net_change
                        total_token_movement += abs(net_change)
            
            # Include address if it meets thresholds or is the sender
            if (
                total_token_movement > 0
                or total_currency_movement > 0
                or addr == from_address
            ):
                net[addr] = {
                    "token_net": token_net,
                    "currency_net": currency_net,
                    "movements": {
                        "tokens": dict(self.movements["tokens"][addr]),
                        "currencies": dict(self.movements["currencies"][addr]),
                    },
                }
        return net

    # ------------------------------------------------------------------ #
    # Convenience entry points
    # ------------------------------------------------------------------ #
    def calculate_address_balance_changes(
        self,
        txn_hash: str,
        from_address: str,
        block_number: int,
        txn_index: int,
        eth_transfers: list,
        erc20_transfers: list,
    ) -> dict:
        """
        Build movements directly from raw transfer lists
        (legacy entry point).
        """
        self.movements = {
            "currencies": defaultdict(lambda: defaultdict(lambda: {"in": OrderedDict(), "out": OrderedDict()})),
            "tokens": defaultdict(lambda: defaultdict(lambda: {"in": OrderedDict(), "out": OrderedDict()})),
        }

        # ERC-20 transfers
        for tr in erc20_transfers:
            tid = (block_number, txn_index, tr["log_index"])
            token_addr = tr["token_address"]
            
            if token_addr in DENOM_ADDRESSES:
                # Known currencies/stablecoins - track as currency with decimal conversion
                currency_symbol = DENOM_ADDRESSES[token_addr]
                
                # Special handling for WETH - always treat as ETH
                if currency_symbol == 'WETH':
                    currency_symbol = 'ETH'
                
                # Apply decimal conversion if known
                if currency_symbol in ERC20_TOKEN_DECIMALS:
                    decimals = ERC20_TOKEN_DECIMALS[currency_symbol]
                    amount = tr["amount"] / (10 ** decimals)
                else:
                    amount = tr["amount"]
                    
                self._track_movement(
                    "currency", tr["from_address"], tr["to_address"], amount, tid, currency=currency_symbol
                )
            else:
                # Unknown tokens - track as regular tokens with raw amounts
                self._track_movement(
                    "token", tr["from_address"], tr["to_address"], tr["amount"], tid, token_address=token_addr
                )

        # ETH (internal) transfers
        for tr in eth_transfers:
            log_index_or_depth = tr.get("log_index") or tr.get('depth')
            tid = (block_number, txn_index, log_index_or_depth)
            self._track_movement(
                "currency", tr["from_address"], tr["to_address"], tr["amount"], tid, currency="ETH"
            )

        try:
            return self.get_net_balance_changes(from_address)
        except Exception as exc:
            if self.logger:
                self.logger.error(
                    f"State diff error for {txn_hash}: {exc}", exc_info=True
                )
            return {}

    def calculate_address_balance_changes_from_processed_tx(self, processed_tx) -> dict:
        """
        Preferred path: use a `ProcessedTransaction` object that already contains
        parsed ERC-20 and internal transfers.
        """
        self.movements = {
            "currencies": defaultdict(lambda: defaultdict(lambda: {"in": OrderedDict(), "out": OrderedDict()})),
            "tokens": defaultdict(lambda: defaultdict(lambda: {"in": OrderedDict(), "out": OrderedDict()})),
        }

        bn, txi = processed_tx.block_number, processed_tx.txn_index

        # 0️⃣ Handle basic transaction value transfer (if any)
        # Skip if there's a depth-0 internal transaction that already captures this transfer
        has_depth_zero_internal = any(
            it.depth == 0 and 
            it.from_address == processed_tx.from_address and 
            it.to_address == processed_tx.to_address and
            abs(float(it.value) - float(processed_tx.value)) < 1e-18  # Use epsilon for float comparison
            for it in processed_tx.internal_transactions
        )
        
        if processed_tx.value > 0 and processed_tx.to_address and not has_depth_zero_internal:
            # Basic ETH transfer from sender to receiver (only if not already captured internally)
            basic_transfer_id = (bn, txi, "basic_transfer")
            self._track_movement(
                "currency", 
                processed_tx.from_address, 
                processed_tx.to_address, 
                # Store ETH amounts in wei to keep units consistent with internal transfers
                float(processed_tx.value) * 1e18,
                basic_transfer_id,
                currency="ETH"
            )

        # 1️⃣ ERC-20 transfers
        for tr in processed_tx.erc20_transfers:
            tid = (bn, txi, tr.log_index)
            token_addr = tr.token_address
            
            if token_addr in DENOM_ADDRESSES:
                # Known currencies/stablecoins - track as currency with decimal conversion
                currency_symbol = DENOM_ADDRESSES[token_addr]
                
                # Special handling for WETH - always treat as ETH
                if currency_symbol == 'WETH':
                    currency_symbol = 'ETH'
                
                # Apply decimal conversion if known
                if currency_symbol in ERC20_TOKEN_DECIMALS:
                    decimals = ERC20_TOKEN_DECIMALS[currency_symbol]
                    amount = float(tr.amount) / (10 ** decimals)
                else:
                    amount = float(tr.amount)
                    
                self._track_movement(
                    "currency", tr.from_address, tr.to_address, amount, tid, currency=currency_symbol
                )
            else:
                # Unknown tokens - track as regular tokens with raw amounts
                self._track_movement(
                    "token", tr.from_address, tr.to_address, float(tr.amount), tid, token_address=token_addr
                )

        # 2️⃣ Internal ETH transfers
        for internal_index, it in enumerate(processed_tx.internal_transactions):
            tid = (bn, txi, f"internal_{internal_index}")  # Use sequential index for uniqueness
            self._track_movement(
                "currency", it.from_address, it.to_address, float(it.value), tid, currency="ETH"
            )

        try:
            return self.get_net_balance_changes(processed_tx.from_address)
        except Exception as exc:
            if self.logger:
                self.logger.error(
                    f"State diff calc failed for {processed_tx.hash}: {exc}",
                    exc_info=True,
                )
            return {}
